compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFoldDmaWaits.cpp

```diff
@@ -62,93 +107,242 @@ FailureOr<bool> canFoldBasedOnHalfDmaCpy(
   uint32_t row = getConstantIndexOrAssert(tileOp.getRow());
   uint32_t maxQueueSize = deviceModel.getDmaMaxQueueSize(col, row);
 
-  // Keep wait op if, either reaches the maximum queue size, or a
-  // duplicate BD ID in the same tile, or packet flow, or the queue is
-  // empty
   uint32_t bdId = getConstantIndexOrAssert(bdIdOp.getValue());
-  bool isDuplicateBdId =
-      llvm::any_of(tileConnectToBdIdQueue, [&](const auto &entry) {
-        return entry.first.first == tileOp &&
-               llvm::is_contained(entry.second, bdId);
-      });
-  SmallVector<uint32_t> &bdIdQueue =
-      tileConnectToBdIdQueue[{tileOp, connectionOp}];
+  bool isDuplicateBdId = llvm::any_of(dmaBdIdsMap, [&](const auto &entry) {
+    return entry.first.first == tileOp && entry.second.contains(bdId);
+  });
+  DenseSet<uint32_t> &bdIds = dmaBdIdsMap[{tileOp, connectionOp}];
   bool canFold = true;
-  if (isDuplicateBdId || isPacketFlow || bdIdQueue.size() >= maxQueueSize ||
-      bdIdQueue.empty()) {
-    bdIdQueue.clear();
+  // Can't fold wait op if:
+  // (1) the current BD ID on the same tile already occurs in the queue, or
+  // (2) the current operation is a packet flow, or
+  // (3) reaches the maximum queue size, or
+  // (4) the queue is empty, or
+  // (5) the current operation is not in the same scope as the queue.
+  if (isDuplicateBdId || isPacketFlow || bdIds.size() >= maxQueueSize ||
+      bdIds.empty() || !isSameScope) {
+    bdIds.clear();
```

Comment:
I know this is pre-existing logic, but try to avoid changing data in `can/is...` functions to make more descriptive and clear of what is going on (not for this PR).

Reply:
will have them in a separate PR